library/src/main/java/io/ona/kujaku/services/TrackingService.java

```diff
@@ -0,0 +1,939 @@
+                    Log.w(TAG,
+                            "Abort service when starting because GPS not enabled.");
+
+                    // Stop the service
+                    stopSelf();
+
+                    return Service.START_NOT_STICKY;
+                }
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+
+        Log.d(TAG, "Tracking service stopping.");
+
+        try {
+            // Remove listeners
+            if (locationManager != null && locationListener != null) {
+                Log.d(TAG, "Remove location manager updates.");
+                locationManager.removeUpdates(locationListener);
+            }
+
+            // Stop the service thread by posting a runnable in the loop.
+            if (gpsHandler != null) {
+                Log.d(TAG, "Quitting looper");
+                gpsHandler.post(stopServiceThread);
+            }
+
+            if (wakeLock != null && wakeLock.isHeld()) {
+                Log.d(TAG, "Release wake lock.");
+                wakeLock.release();
+            }
+
+        } catch (IllegalArgumentException e) {
+            Log.e(TAG, "Failed to stop service properly.", e);
+        }
+
+        Log.d(TAG, "Wait for the threads to exit.");
+
+        // Wait for the threads to die. This is required to implement an async stop. See Utils.
+        try {
+            if (serviceThreadRunningLatch != null) {
+                if (!serviceThreadRunningLatch.await(WAIT_TIME_SERVICE_THREAD, TimeUnit.MILLISECONDS)) {
+                    Log.w(TAG, "Time out waiting for service thread to exit.");
+                }
+                Log.d(TAG, "Service thread has stopped.");
+            }
+
+        } catch (InterruptedException ie) {
+            Log.e(TAG, "Main application thread was interrupted.", ie);
+        }
+
+        setServiceStatus(TrackingServiceStatus.STOPPED);
+
+        super.onDestroy();
+
+        Log.i(TAG, "Tracking service stopped.");
+    }
+
+    @Override
+    public IBinder onBind(Intent intent) {
+        return binder;
+    }
+
+    /**
+     *
+     *
+     * @param intent
+     */
+    private void createNotificationPendingIntent(Intent intent) {
+        Class<?> cls = getActivityClassFromCanonicalName(intent);
+
+        // The stack builder object will contain an artificial back stack for the started Activity.
+        // This ensures that navigating backward from the Activity leads out of
+        // your application to the Home screen.
+        // ALL THIS is required to start the service in the foreground!
+        TaskStackBuilder stackBuilder = TaskStackBuilder.create(this);
+
+        // Adds the back stack for the Intent (but not the Intent itself)
+        if (cls != null) {
+            // Creates an explicit intent for an Activity
+            Intent startActivityIntent = new Intent(this, cls)
+                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+
+            stackBuilder.addParentStack(cls);
```

Comment:
What is the aim of this in this code? Not sure of the purpose since `FLAG_ACTIVITY_CLEAR_TASK` and `FLAG_ACTIVITY_NEW_TASK` kinda does this as according to this

```
/**
     * If set in an Intent passed to {@link Context#startActivity Context.startActivity()},
     * this flag will cause any existing task that would be associated with the
     * activity to be cleared before the activity is started.  That is, the activity
     * becomes the new root of an otherwise empty task, and any old activities
     * are finished.  This can only be used in conjunction with {@link #FLAG_ACTIVITY_NEW_TASK}.
     */
```

Do they do the same thing for now or did you have a different usage for the client apps?

Reply:
Honestly i'm not really an expert on this. So i did some tests to understand how it works. I definitively agree that we don't need those flags

Reply:
Yea, so we can either do away with the `TaskStackBuilder` or remove the flags

Reply:
Fixed